lib/Pipes/PipelineManager.cpp

```diff
@@ -448,3 +448,35 @@ PipelineManager::invalidateFromDiff(const llvm::StringRef Name,
   // TODO: once invalidations are working, return `Map` instead of this
   return invalidateAllPossibleTargets();
 }
+
+llvm::Error PipelineManager::produceTargets(const llvm::StringRef StepName,
+                                            const Container &TheContainer,
+                                            const TargetsList &List,
+                                            llvm::raw_ostream &Output) {
+  if (CurrentState.count(StepName) == 0)
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "Step not found");
+
+  if (!CurrentState[StepName].contains(TheContainer.first()))
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "Container not found");
+
+  auto &CurrentContainerState = CurrentState[StepName].at(TheContainer.first());
+  for (const pipeline::Target &Target : List) {
+    if (!CurrentContainerState.contains(Target))
+      return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                     "Target not found");
+  }
+
+  ContainerToTargetsMap Targets;
+  for (const pipeline::Target &Target : List)
+    Targets[TheContainer.second->name()].push_back(Target);
+
+  if (auto Error = getRunner().run(StepName, Targets); Error)
+    return Error;
+
+  const auto &ToFilter = Targets.at(TheContainer.second->name());
+  const auto &Cloned = TheContainer.second->cloneFiltered(ToFilter);
+  llvm::cantFail(Cloned->serialize(Output));
```

Comment:
i would keep this part of the function in the pipeline C, there is no reason why the production routine should have a opinion regarding how to return what has been requested, in particular if it returns them by serializing them into a stream

Reply:
As discussed IRL, split the functionality in `materializeTargets` and `produceTargets`